src/audio/dai.c

```diff
@@ -635,6 +691,57 @@ static int dai_comp_trigger(struct comp_dev *dev, int cmd)
 	return ret;
 }
 
+static int dai_comp_trigger(struct comp_dev *dev, int cmd)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+	struct dai_group *group = dd->group;
+	uint32_t irq_flags;
+	int ret = 0;
+
+	if (!group) {
+		comp_dbg(dev, "dai_comp_trigger(), non-atomic trigger");
+		return dai_comp_trigger_int(dev, cmd);
+	}
+
+	if (group->trigger_cmd < 0) {
+		comp_dbg(dev, "dai_comp_trigger(), begin atomic trigger for group %d",
+			 group->group_id);
+		group->trigger_cmd = cmd;
+		group->trigger_counter = group->num_dais - 1;
+	} else if (group->trigger_cmd != cmd) {
+		comp_err(dev, "dai_comp_trigger(), already processing atomic trigger");
+		/* already processing a different trigger command */
+		ret = -EAGAIN;
+	} else {
+		group->trigger_counter--;
+		comp_dbg(dev, "dai_comp_trigger(), trigger counter %d, group %d",
+			 group->trigger_counter, group->group_id);
+
+		if (!group->trigger_counter) {
+			irq_local_disable(irq_flags);
+			notifier_event(group, NOTIFIER_ID_DAI_TRIGGER,
+				       BIT(cpu_get_id()), NULL, 0);
+			irq_local_enable(irq_flags);
+
+			ret = group->trigger_ret;
+			group->trigger_cmd = -1;
+			group->trigger_counter = 0;
+		}
+	}
+
+	return ret;
+}
+
+static void dai_atomic_trigger(void *arg, enum notify_id type, void *data)
+{
+	struct comp_dev *dev = arg;
+	struct dai_data *dd = comp_get_drvdata(dev);
+	int ret = dai_comp_trigger_int(dev, dd->group->trigger_cmd);
+
+	if (ret && !dd->group->trigger_ret)
```

Comment:
i didn't get the logic here, what is this trying to test?

Reply:
It keeps the first reported trigger error. If there's already an error stored in the field, it won't overwrite it.

Reply:
I guess I'm gonna change it to last reported error instead. It shouldn't matter and it will let me remove that if statement.